ponzu/llama/curatedTVL.py

```diff
@@ -0,0 +1,130 @@
+        return self.raw_tvl
+
+    def _reclassify_tvl(self):
+        self.raw_tvl.loc[self.raw_tvl['protocol'] == 'eos-rex', 'category'] = 'Liquid Staking'
+        self.raw_tvl.loc[self.raw_tvl['protocol'] == 'atomichub', 'category'] = 'NFT Marketplace'
+
+    def curate_tvl(self, chains):
+        curated_chains = ['ethereum','tron','binance','arbitrum','polygon','optimism','avalanche','fantom','solana','cardano','eos','tezos','aptos','near','stacks','sui','hedera','aurora','eos evm','wax','europa']
+        for chain in chains:
+            if chain not in curated_chains:
+                return f"Error: {chain} is not in the curated list of chains." 
+        if self.raw_tvl is None:
+            self.fetch_tvl(chains)
+
+        self._filter_tvl()
+        return self.curated_tvl
+
+    def _filter_tvl(self):
+        df = self.raw_tvl[(self.raw_tvl['type'] == 'tvl') | (self.raw_tvl['type'] == 'borrowed')]
+        exclude_categories = ['CEX', 'Chain', 'Yield', 'Bridge', 'Liquid Staking', 'Liquidity manager', 'RWA', 'Gaming', 'NFT Marketplace', 'Leveraged Farming', 'Prediction Market']
+        df = df[~df['category'].isin(exclude_categories)]
+        exclude_protocols = ['ribbon','frax','unslashed','sturdy','alongside','illuminate','instadapp','defisaver','pooltogether','cian','ubiquity-dao','sandclock','templedao','unbound','axia-protocol','tornado-cash','aztec','tronnrg','garble.money','sperax-demeter','volta-finance','uniwswap-unia-farms','mm-stableswap-polygon','zkbob','brokkr-finance','defrost','sherpa-cash','sturdy','shadecash','zeta']
+        df = df[~df['protocol'].isin(exclude_protocols)]
+        self.curated_tvl = df
+        return df
+    
+    def _prepare_tvl(self, chains, curated):
+        if curated:
+            self.tvl = self.curate_tvl(chains)
+        else:
+            self.tvl = self.fetch_tvl(chains)
+
+    def _get_pivot_table(self, values, index, columns, aggfunc, tvl=None):
+        if tvl is None:
+            tvl = self.tvl
+        return pd.pivot_table(tvl, values=values, index=index, columns=columns, aggfunc=aggfunc)
+
+    def chain_tvl(self, chains, curated=False):
+        self._prepare_tvl(chains, curated)
+        return self._get_pivot_table('tvl', 'date', 'chain', sum)
+
+    def chain_protocols_tvl(self, chains, curated=False):
+        self._prepare_tvl(chains, curated)
+        tvl_protocols = {}
+        for chain in chains:
+            tvl_protocols_chain = self.tvl[self.tvl['chain'] == chain]
+            tvl_protocols[f'tvl_protocols_{chain}'] = pd.pivot_table(tvl_protocols_chain, values='tvl', index='date', columns='protocol', aggfunc=sum)
+
+        return tvl_protocols
+
+    def chain_tvb(self, chains, curated=False):
+        self._prepare_tvl(chains, curated)
+        df_borrows = self.tvl[self.tvl['type'] == 'borrowed']
+        return self._get_pivot_table('tvl', 'date', 'chain', sum, df_borrows)
+
+    def chain_protocols_tvb(self, chains, curated=False):
```

Comment:
what is the point of this fn? you are just making filtered tables stored in a dict -- why not just do that in one line of code when you get the data? 

like how exactly are you using this output?

Reply:
certainly could, but if I'm doing that a bunch of times (i.e. using it for L1 report and each individual quarterly reports) as well as sharing it with ppl who maybe don't know how to filter a df then would rather just bake into fn. what is the downside?

Reply:
bc the two use cases i can think of are a) plotting multiple chains on a chart or b) plotting one of the chains in the df 

so in case a) you would have to re put all these back together and should just do a single group by when you consume it instead and in case b) why not just call the function for the specific chain 

guess i'm just thinking of how i would use the result and feels funky but i would need to see how you are planning on using this downstream

Reply:
if a) would call the chain fn and not the chain_protocols fn
b) can call raw function for a chain but then will have to do a line or two of code. idea with these is to be able to format data in the way I use it every time by just calling fn. I agree they are not doing much, but do they hurt to have?